experimental/iree_profiler/profiler.py

```diff
@@ -0,0 +1,135 @@
+                      "if .vmfb file is present.")
+  parser.add_argument("--compile-only", default='False', \
+                      type=str, help="Compiles the operation "\
+                        "without running verification and profiling.")
+
+  # Profiling options
+  parser.add_argument("--profiling-enabled", "--benchmark", default='True', \
+                      type=str, help="Benchmark the operation.")
+  parser.add_argument('--batch-size', '--benchmark-dispatch-repeat-count', \
+                      default=100, help="Number of times dispatch is launched "\
+                        "in a loop to amortize the launch overhead.")
+  parser.add_argument("--benchmark-repetitions", default=5,
+                      type=int, help="Number of times benchmark is repeated "\
+                      "and min, max, median, and average runtimes/gflops are "\
+                      "reported.")
+
+  # Verification options
+  parser.add_argument("--verification-enabled", default='True',
+                      type=str, help="Verify the operation against reference numpy "\
+                      "implementation.")
+
+  # Performance reporting options
+  parser.add_argument("--output", default='', \
+                      help="Path to output file for csv readable results.")
+  parser.add_argument("--append", default='false', \
+                      help="If true, result is appended to possibly existing file. "\
+                        "Otherwise, any existing file is overwritten.")
+
+  parser.add_argument("--tags", default='', \
+                      help="Inserts leading columns in output table and uniform "\
+                        "values for each column. Useful for generating pivot tables.")
+  args = parser.parse_args()
+  ###############################################################################
+
+  # Boolenize the string arguments from command line.
+  verification_enabled = False if args.verification_enabled in [
+      'False', 'false', '0'
+  ] else True
+  profiling_enabled = False if args.profiling_enabled in [
+      'False', 'false', '0'
+  ] else True
+  compile_only = False if args.compile_only in ['False', 'false', '0'] else True
+  # Overrite verification and profiling if compile_only is set.
+  if compile_only:
+    verification_enabled = False
+    profiling_enabled = False
+
+  # Path to the directory where the generated operations are stored.
+  generated_path = os.path.join(args.build_dir, 'generated', args.mlir_dialect)
+
+  # Manifests metadata for a group of accompanying opeartions and configurations.
+  manifest = Manifest(args)
+
+  # Collect all the avialable operations in a manifest.
+  GpuMatmulTensorCoresF16(manifest)
+  #GpuMatmulTensorCoresF32(manifest)
+
+  # Performance report
+  perf_report = PerformanceReport(args)
+
+  # For all the operations in the manifest, compile and profile them.
+  for operation_kind, operation_collection_list in manifest.operations.items():
```

Comment:
I'm eager to see this tool generalized to profile dispatch regions generally -- not just single operations. While we don't have to go that far in this PR, there are many cases like the use of "operation" in this file, where we could use a general term... not sure if `dispatch_region` would be appropriate, or some other term would be better?

Reply:
> Looks like a great start! I'll look more at the code in a bit but we'll want someone with some python experience to review the basics/style.
> 
> We'll want to rename this, though, to avoid confusion and ensure only the right people find this - it's not an "IREE profiler" as most end users would imagine it: I'd expect an IREE profiler to profile my model and tell me what's slow (and we'll have a tool that does that at some point). This is more a "dispatch evaluator" or something. The closest conceptual equivalent on the LLVM side is llvm-exegesis - maybe this is iree-exegesis? https://llvm.org/docs/CommandGuide/llvm-exegesis.html

Let us decide a top-level directory name and get this fist version merged, Thus others can start using it as well. 

How about `iree/experimental/dispatch-profiler` ? 

cc: @benvanik , @julianwa , @mattwalsh , @stellaraccident name suggestions for the top-level directory

Reply:
`dispatch_profiler` works for me!

> In the code, dispatch_region or operation both are fine. operations is an mlir::Operation that we are using create, compile, and launch a dispatch. To begin with, dispatches I am compiling, verifying, and profiling are linalg.matmul, but could be a combination of operations.

It sounds like using the term `dispatch` would be appropriate here. Even if technically `dispatch` maps to an `mlir::Operation`, `dispatch` has less chance of confusion, and matches the `dispatch_profiler` name of the tool.

Reply:
I made the change to dispatch-profiler. 

Also, gave a lot thought to what should we call operation and what should we call dispatch s.t. improving this work in a collaborative doesn't lead to confusion. Thus, in the code, now we use `operation`, configuration, and`dispatch`  based on definitions below:

```
# The profiler tool is organized as follows:
# Operation: A MLIR operation that is generated or consumed by the
#       dispatch_profiler. For example, linalg.matmul, linalg.conv2d, etc.
# Configuration: A set of compile parameters that are used by iree-compile the
#       to choose a compilation pipeline (e.g. LLVMGPUTensorCore,
#       LLVMGPUTensorCoreMmaSync, LLVGPUCPU, etc.), performance tuning parameters
#       (e.g. workgroup size, tile size etc.).
# Dispatch: A combination of an operation and a configuration is launched by the
#       dispatch profiler for verification and performance profiling. Note that
#       a dispatch is not a MLIR operation it is binary executable that is launched
#       by the profiler. Additionally, the goal of the tool is to also profile the
#       performance of the fusions and a dispatch for fusion is a combination of
#       multiple operations glued together and compiled into a single dispatch.
```